Flaky test in ReceivesTreeUpdatesHandlerTest reveals a race-condition

```
 INFO ide_ci::program::command: sbtℹ️ [info] ReceivesTreeUpdatesHandler
Error: NFO ide_ci::program::command: sbtℹ️ [ERROR] [2023-10-25 18:01:39,754] [org.enso.languageserver.search.SuggestionsHandler] Cannot read the package definition from [/tmp/14189155674160513259].
 INFO ide_ci::program::command: sbtℹ️ [info] - must acquire capability receivesTreeUpdates
 INFO ide_ci::program::command: sbtℹ️ [info] - must fail to acquire capability if directory doesn't exist
 INFO ide_ci::program::command: sbtℹ️ [info] - must reacquire capability receivesTreeUpdates
 INFO ide_ci::program::command: sbtℹ️ [info] - must fail to release capability it does not hold *** FAILED ***
 INFO ide_ci::program::command: sbtℹ️ [info]   java.lang.AssertionError: assertion failed: timeout (5 seconds) during expectMsgClass waiting for class java.lang.String
 INFO ide_ci::program::command: sbtℹ️ [info]   at scala.Predef$.assert(Predef.scala:279)
 INFO ide_ci::program::command: sbtℹ️ [info]   at akka.testkit.TestKitBase.expectMsgClass_internal(TestKit.scala:571)
 INFO ide_ci::program::command: sbtℹ️ [info]   at akka.testkit.TestKitBase.expectMsgClass(TestKit.scala:567)
 INFO ide_ci::program::command: sbtℹ️ [info]   at akka.testkit.TestKitBase.expectMsgClass$(TestKit.scala:567)
 INFO ide_ci::program::command: sbtℹ️ [info]   at akka.testkit.TestKit.expectMsgClass(TestKit.scala:973)
 INFO ide_ci::program::command: sbtℹ️ [info]   at org.enso.jsonrpc.test.JsonRpcServerTestKit$WsTestClient.expectMessage(JsonRpcServerTestKit.scala:117)
 INFO ide_ci::program::command: sbtℹ️ [info]   at org.enso.jsonrpc.test.JsonRpcServerTestKit$WsTestClient.expectJson(JsonRpcServerTestKit.scala:126)
 INFO ide_ci::program::command: sbtℹ️ [info]   at org.enso.languageserver.websocket.json.ReceivesTreeUpdatesHandlerTest.$anonfun$new$5(ReceivesTreeUpdatesHandlerTest.scala:49)
 INFO ide_ci::program::command: sbtℹ️ [info]   at org.scalatest.OutcomeOf.outcomeOf(OutcomeOf.scala:85)
 INFO ide_ci::program::command: sbtℹ️ [info]   at org.scalatest.OutcomeOf.outcomeOf$(OutcomeOf.scala:83)
 INFO ide_ci::program::command: sbtℹ️ [info]   ...
 ```
 
 https://github.com/enso-org/enso/actions/runs/6642894609/job/18048711561?pr=8145
 
I encountered the problem many times, this is a recent one.
After a bit more digging I figured that there is a race-condition between the `PathWatcher` handler sending a reply and it terminating and therefore not unregistering from `ReceivesTreeUpdatesHandler` store.

--- a/engine/language-server/src/main/scala/org/enso/languageserver/filemanager/PathWatcher.scala
+++ b/engine/language-server/src/main/scala/org/enso/languageserver/filemanager/PathWatcher.scala
@@ -18,11 +18,14 @@ import org.enso.languageserver.data.{
 }
 import org.enso.languageserver.effect._
 import org.enso.languageserver.event.JsonSessionTerminated
+import org.enso.languageserver.filemanager.PathWatcher.{
+  ForwardRequest,
+  ForwardResponse
+}
 import org.enso.languageserver.util.UnhandledLogging
 import zio._
 
 import java.io.File
-
 import scala.concurrent.Await
 import scala.util.{Failure, Success}
 
@@ -108,9 +111,14 @@ final class PathWatcher(
       sender() ! CapabilityAcquired
       context.become(initializedStage(root, base, clients ++ newClients))
 
-    case UnwatchPath(client) =>
-      sender() ! CapabilityReleased
-      unregisterClient(root, base, clients - client)
+    case ForwardRequest(originalSender, UnwatchPath(client)) =>
+      val newClients = clients - client
+      sender() ! ForwardResponse(
+        originalSender,
+        CapabilityReleased,
+        newClients.isEmpty
+      )
+      unregisterClient(root, base, newClients)
 
     case JsonSessionTerminated(client)
         if clients.contains(client.rpcController) =>
@@ -222,6 +230,21 @@ object PathWatcher {
     def canRestart: Boolean =
       restartCount < maxRestarts
   }
+
+  /** A PathWatcher request that should be replied to the handler rather than sender.
+    *
+    * @param sender the original sender of the request
+    * @param request the request to be handled by `PathWatcher`
+    */
+  case class ForwardRequest(sender: ActorRef, request: Any)
+
+  /** A PathWatcher response to the proxy (handler) containing information about the original sender.
+    *
+    * @param sender the original sender of the request
+    * @param response response that should be sent to the `sender`
+    * @param last true, if the handling path watcher will shutdown after serving this request, false otherwise
+    */
+  case class ForwardResponse(sender: ActorRef, response: Any, last: Boolean)
 
   /** Creates a configuration object used to create a [[PathWatcher]].
     *

--- a/engine/language-server/src/main/scala/org/enso/languageserver/filemanager/ReceivesTreeUpdatesHandler.scala
+++ b/engine/language-server/src/main/scala/org/enso/languageserver/filemanager/ReceivesTreeUpdatesHandler.scala
@@ -15,6 +15,10 @@ import org.enso.languageserver.data.{
 }
 import org.enso.languageserver.effect._
 import org.enso.languageserver.event.InitializedEvent
+import org.enso.languageserver.filemanager.PathWatcher.{
+  ForwardRequest,
+  ForwardResponse
+}
 import org.enso.languageserver.util.UnhandledLogging
 
 /** Handles `receivesTreeUpdates` capabilities acquisition and release.
@@ -146,9 +150,18 @@ final class ReceivesTreeUpdatesHandler(
         ) =>
       store.getWatcher(path) match {
         case Some(watcher) =>
-          watcher.forward(PathWatcherProtocol.UnwatchPath(client.rpcController))
+          watcher ! ForwardRequest(
+            sender(),
+            PathWatcherProtocol.UnwatchPath(client.rpcController)
+          )
         case None =>
           sender() ! CapabilityNotAcquiredResponse
+      }
+
+    case ForwardResponse(client, response, isLast) =>
+      client.forward(response)
+      if (isLast) {
+        context.become(withStore(store.removeWatcher(sender())))
       }
 
     case Terminated(watcher) =>

